### PR TITLE
Add inner and outer stair corners

### PR DESCRIPTION
**Pull request by MarkuBu**
This PR adds simple stairs for corners. It uses nodeboxes. 

If someone can create mesh nodes I will add them

### CODE REVIEW

**Comment by MarkuBu** [comment]
https://github.com/minetest/minetest_game/issues/1443

**Comment by paramat** [comment]
These will have a lot of hidden faces so meshnodes are needed for these. Good feature though.

**Comment by MarkuBu** [comment]
I know that nodeboxes are suboptimal. Will change that as soon as I have mesh nodes. But I can't make them myself because I have absolutely no Blender skills.

Recipes are added, but not uploaded

But I don't know what you mean with _the cut-away part is diagonally facing the player_

**Comment by paramat** [comment]
68747470733a2f2f7331372e706f7374696d672e6f72672f67386f68736f6c6e6a2f636f726e65725f7374616972732e706e67

See the orientations on the right, that is how the node should be placed relative to the player facing diagonally.

**Comment by paramat** [comment]
Then maybe also use the existing 'pointed thing to face pos' function to turn the corner stair upside-down depending on what part of a wall's tile is pointed at.

**Comment by paramat** [comment]
The meshnodes can be swapped in much later when they're done, maybe even in another PR, nodeboxes are fine for now while this is developed.

**Comment by MarkuBu** [comment]
I add rotate and place. I think it works intuitive.

File: mods/stairs/init.lua [hunk 1]
```diff
@@ -255,13 +255,137 @@ if replace then
...
+		local baseburntime = minetest.get_craft_result({
+			method = "fuel",
+			width = 1,
+			items = {recipeitem}
+		}).time
+		if baseburntime > 0 then
+			minetest.register_craft({
+				type = "fuel",
+				recipe = 'stairs:stair_inner_' .. subname,
+				burntime = math.floor(baseburntime * 0.75),
```
Comment (paramat) [inline mods/stairs/init.lua]: 5 small cubes out of 8:
`baseburntime * 0.625`

File: mods/stairs/init.lua [hunk 2]
```diff
@@ -255,13 +255,137 @@ if replace then
...
+		local baseburntime = minetest.get_craft_result({
+			method = "fuel",
+			width = 1,
+			items = {recipeitem}
+		}).time
+		if baseburntime > 0 then
+			minetest.register_craft({
+				type = "fuel",
+				recipe = 'stairs:stair_outer_' .. subname,
+				burntime = math.floor(baseburntime * 0.75),
```
Comment (paramat) [inline mods/stairs/init.lua]: 7 small cubes out of 8:
`baseburntime * 0.875`

**Comment by paramat** [comment]
I'll test sometime.

File: mods/stairs/init.lua (hunk 1 above)
Comment (MarkuBu) [inline mods/stairs/init.lua]: Ah, now I understand that. Will change tomorrow

File: mods/stairs/init.lua [hunk 3]
```diff
@@ -255,13 +255,137 @@ if replace then
...
+		on_place = function(itemstack, placer, pointed_thing)
+			if pointed_thing.type ~= "node" then
+				return itemstack
+			end
+
+			return rotate_and_place(itemstack, placer, pointed_thing)
+		end,
+	})
+
+		if recipeitem then
```
Comment (SmallJoker) [inline mods/stairs/init.lua]: Please repair these indents. The following lines should have the same level as `local baseburntime`.

Comment (MarkuBu) [inline mods/stairs/init.lua]: you're right. No idea why there was an indention. Fixed in the next push

**Comment by GreenXenith** [comment]
Is this PR still needing meshes? If the meshes are done before June 4th will this be merged for 0.4.16 release?

**Comment by GreenXenith** [comment]
@MarkuBu here's some meshes:
[corner_stairs.zip](https://github.com/minetest/minetest_game/files/1033167/corner_stairs.zip)

**Comment by TumeniNodes** [comment]
he was looking for .b3d, not sure why it matters. .obj should work fine but if you are able to export .b3d to add, I think he would be happy

sorry to chime in,

**Comment by GreenXenith** [comment]
Don't know why he would use .b3d if the normal stair is .obj...

**Comment by TumeniNodes** [comment]
no idea. .obj is pretty much the standard for MT. .b3d offers better animation but def not needed in this instance. I think at this point he will be happy with these. esp if it will mean being merged now : )

**Comment by MarkuBu** [comment]
I'm sorry. Cause I work with mobs I had b3d in mind. obj is fine too.

But I don't have Internet on my PC I can't make any changes right now

**Comment by GreenXenith** [comment]
Good thing you can't do it right now. The models need to be fixed anyway :/
Fixing now...

**Comment by paramat** [comment]
Should be .obj like the stair.
This will not be in release because 'feature freeze', bugfixes only.

**Comment by GreenXenith** [comment]
Problem. The wooden outer stair has the wrong wood grain on the top part on the right side, but if I fix that, the bronze outer stair (or any metal block for that matter) will have the wrong outline orientation.
screenshot_20170527_112225

**Comment by MarkuBu** [comment]
I'm temporally back in business, but if this won't be add to 0.4.16 there is no hurry

**Comment by paramat** [comment]
It should be possible to fix both, normally a texture is projected onto the side so the top half of the wood stair should not be vertical anyway.

**Comment by GreenXenith** [comment]
@MarkuBu here are working meshes. Note, you will have to define new hit/selection boxes.
[corner_stairs.zip](https://github.com/minetest/minetest_game/files/1033765/corner_stairs.zip)

**Comment by TumeniNodes** [comment]
If it could be considered that stairs are missing the outer / inner corners, is indeed a "bug", this becomes a bugfix.
(could be a sleazy, but workable way of working this into release)? Knowing this addition will not add any actual bugs. And would be a very welcome addition to end users.

**Comment by MarkuBu** [comment]
@GreenXenith thanks, I try it tomorrow. They have to behave like the nodebox version without changing the code to rotate them

**Comment by paramat** [comment]
About nodeboxes that get converted to meshes, they get generated with the internal faces of all boxes, which is why .obj meshes are better, no internals, faster render.
This is why slabs are not .obj, no internal faces so no advantage to using a .obj.

**Comment by paramat** [comment]
MarkuBu, these will need a dedicated placement function because these are diagonally-orientated nodes, so their rotation should switch at player yaw 0, 90, 280, 270 degrees instead of 45, 135, 225, 315 degrees. The function can use the same 'pointed thing to face pos' code for deciding upright or upside-down.

**Comment by paramat** [comment]
Also .obj meshes are needed.

**Comment by paramat** [comment]
@MarkuBu are you able to finish this? A new PR has appeared #1814 but it has no meshnodes.

**Comment by MarkuBu** [comment]
@paramat as I mentioned in the forum I reduce my Minetest activity because I want to start my own project  and I don't have time for both. 

https://forum.minetest.net/viewtopic.php?f=3&t=17879

**Comment by paramat** [comment]
Ok. You've been a good contributor and will be missed.

New PR #1820

**Pull request closed without merging**